### DISCUSSION TITLE
How to extend the StatefulComponentProps type

### DISCUSSION THREAD

**Discussion by convers39** [visgl/deck.gl#7609] [Q&A] [answered]
I am trying to add some custom props when instantiating a layer, which generates the tooltip text for the specific layer.

The issue is when trying to use in `getTooltip` of `DeckGL`, it shows that my custom property does not exist on `StatefulComponentProps`.

Is there any workaround to solve this issue?

<img width="770" alt="image" src="https://user-images.githubusercontent.com/42597395/215636448-29ee117c-8496-4398-a3cc-9839e0461f5d.png">

<img width="1172" alt="image" src="https://user-images.githubusercontent.com/42597395/215636317-00958f8d-5883-470e-9650-88055c054e6d.png">

**Comment by Pessimistress** [comment, marked as answer]
```typescript
type ExtraPropsT = {
  getTooltipText: ...
}

new TripsLayer<Polyline, ExtraPropsT>(...)
```

**Reply by convers39** [reply]
Thank you for your reply, I tried with the code above and still face the same problem. 

<img width="1170" alt="image" src="https://user-images.githubusercontent.com/42597395/215653805-3736b315-0ece-438d-9e61-3b07bd4aa364.png">

Is there any way to extend the type of `StatefulComponentProps`, as even if `new TripsLayer<Polyline, ExtraPropsT>` works I have to add the same `ExtraPropsT` to all the layers?

**Reply by Pessimistress** [reply]
`(layer as TripsLayer<Polyline, ExtraPropsT>).props`

**Reply by convers39** [reply]
Thank you! I wanted it to be a general solution so tried the code below which resolved the ts error.

```ts
import { Layer } from '@deck.gl/core/typed';

type ExtraProps = {
  getTooltipText: (data: unknown) => string;
};

// ...
      <DeckGL
        onViewStateChange={params => {
          updateViewState(params.viewState as MapViewState);
        }}
        viewState={viewState}
        layers={layers}
        controller
        getTooltip={({ layer, object }) => {
          if (layer && object && 'getTooltipText' in layer.props) {
            return (layer as Layer<ExtraProps>).props.getTooltipText(object);
          }
          return null;
        }}
      >
        <Map
          mapStyle={`mapbox://styles/mapbox/${mapStyle}`}
          mapboxAccessToken={MAPBOX_ACCESS_TOKEN}
        />
      </DeckGL>

```
